### PR TITLE
python: Add runnable unittest tasks

### PR DESCRIPTION
**Pull request by rayduck**
Add runnable tasks for Python, starting with `unittest` from the standard library. Both `TestCase`s (classes meant to be a unit of testing) and individual test functions in a `TestCase` will have runnable icons. For completeness, I also included a task that will run `unittest` on the current file.

The implementation follows the `unittest` CLI. The unittest module can be used from the command line to run tests from modules, classes or even individual test methods:

```
python -m unittest test_module.TestClass
python -m unittest test_module.TestClass.test_method
```

```python
import unittest

class TestStringMethods(unittest.TestCase):

    def test_upper(self):
        self.assertEqual('foo'.upper(), 'FOO')

    def test_isupper(self):
        self.assertTrue('FOO'.isupper())
        self.assertFalse('Foo'.isupper())

    def test_split(self):
        s = 'hello world'
        self.assertEqual(s.split(), ['hello', 'world'])
        # check that s.split fails when the separator is not a string
        with self.assertRaises(TypeError):
            s.split(2)

if __name__ == '__main__':
    unittest.main()
```

From the snippet provided by `unittest` docs, a user may want to run test_split independently of the other test functions in the test case. Hence, I decided to make each test function runnable despite `TestCase`s being the unit of testing.

## Example of running a `TestCase`
image

## Example of running a test function in a `TestCase`
image

`unittest` will also run the `setUp` and `tearDown` fixtures.

Eventually, I want to add the more commonly used `pytest` runnables (perhaps as an extension instead). 

Release Notes:

- Added runnable tasks for Python `unittest`. ([#12080](https://github.com/zed-industries/zed/issues/12080)).

### CODE REVIEW

File: crates/languages/src/python.rs [hunk 1]
```diff
@@ -196,6 +196,42 @@ pub(super) fn python_task_context() -> ContextProviderWithTasks {
...
+            args: vec![
+                "-m".to_owned(),
+                "unittest".to_owned(),
+                VariableName::RelativeFile.template_value(),
+                "-k".to_owned(),
+                "$ZED_SYMBOL".to_owned(),
+            ],
+            tags: vec!["python-unittest-function".to_owned()],
+            ..TaskTemplate::default()
+        },
```
Comment (rayduck) [inline crates/languages/src/python.rs:234]: This task for functions looks the same as the that for the `TestCase`. Ideally, `ZED_SYMBOL` should be something like `test_module.TestClass.test_method` which enables precisely targetting the test to run. Currently, if there are two identically named test functions (from different `TestCase` classes) in the same file, both will be run when one of them is pressed, since I'm filtering by file name and function name.

I am not sure if this is a current limitation or that there is some way to construct this string from tree-sitter queries (since I know both the node names of the TestCase class and the name of the function). @osiewicz perhaps you could provide some pointers here?

Comment (osiewicz) [reply crates/languages/src/python.rs:234]: Hey, thanks for submitting this PR; the answer to your woes is: ZED_SYMBOL is probably not what you want to use as extension author. It was one of the first task variables we've exposed and it's quite limited in what it can do (which you pointed out already). 
Yes, it is possible to construct your own environmental variables to expose for tasks. Long story short: any capture not prefixed with `_` will be set in the environment of the task you're spawning under the `ZED_CUSTOM_{CAPTURE_NAME}` name. E.g: a full text of a capture named `food` is stored in `ZED_CUSTOM_food`.
Now, you want to do some processing on the values of the captures as well, right? You want to concatenate them with `.` delimiter. You can do that too. Built in languages (which Python classifies as) can inspect variables from captures and add arbitrary environment variables. Note that they can inspect all variables from captures, including those with `_` prefix.
TL;DR: you can take a look at how Go does it: https://github.com/zed-industries/zed/blob/main/crates/languages/src/go.rs#L453
It implements `ContextProvider` trait which has a `build_context` method. This method takes `task_variables`, which contain the values of captures. It then returns variables which should be appended to that context.

Comment (osiewicz) [reply crates/languages/src/python.rs:234]: I went ahead and created a patch that will allow you to focus on implementing `build_context` instead of wrestling with setting it all up. 
[your_own_context_provider_in_python.patch](https://github.com/user-attachments/files/15521749/your_own_context_provider_in_python.patch)
Give that a try. Pro tip: you can run a task that spawns `env` (it can even be a one-shot task) anywhere in the code to see what environmental variables are available for you at a given point.

File: crates/languages/src/python.rs [hunk 2]
```diff
@@ -182,57 +182,94 @@ async fn get_cached_server_binary(
...
+            "_unittest_method_name",
+        )));
+
+        let _unittest_target_str = match (_unittest_class_name, _unittest_method_name) {
+            (Some(class_name), Some(method_name)) => {
+                format!("{}.{}.{}", _python_module_name, class_name, method_name)
+            }
+            (Some(class_name), None) => format!("{}.{}", _python_module_name, class_name),
+            (None, None) => format!("{}", _python_module_name),
+            (None, Some(_)) => unreachable!("Method name without class name is not possible"),
```
Comment (osiewicz) [inline crates/languages/src/python.rs:212]: we probably shouldn't crash the whole app, and just return an `Ok(TaskVariables::default())` instead.

File: crates/languages/src/python.rs [hunk 3]
```diff
@@ -182,57 +182,94 @@ async fn get_cached_server_binary(
...
+    VariableName::Custom(Cow::Borrowed("PYTHON_UNITTEST_TARGET"));
+
+impl ContextProvider for PythonContextProvider {
+    fn build_context(
+        &self,
+        _variables: &task::TaskVariables,
+        _location: &project::Location,
+        _cx: &mut gpui::AppContext,
+    ) -> Result<task::TaskVariables> {
+        let _python_module_name = python_module_name_from_relative_path(
```
Comment (osiewicz) [inline crates/languages/src/python.rs:197]: I don't think we need to prefix Rust variables with `_`? They're clearly being used.

File: crates/languages/src/python.rs [hunk 4]
```diff
@@ -182,21 +182,94 @@ async fn get_cached_server_binary(
...
+        let _unittest_method_name = _variables.get(&VariableName::Custom(Cow::Borrowed(
+            "_unittest_method_name",
+        )));
+
+        let _unittest_target_str = match (_unittest_class_name, _unittest_method_name) {
+            (Some(class_name), Some(method_name)) => {
+                format!("{}.{}.{}", _python_module_name, class_name, method_name)
+            }
+            (Some(class_name), None) => format!("{}.{}", _python_module_name, class_name),
+            (None, None) => format!("{}", _python_module_name),
```
Comment (osiewicz) [inline crates/languages/src/python.rs:211]:
```suggestion
            (None, None) => _python_module_name,
```

Comment (rayduck) [reply crates/languages/src/python.rs:211]: Thanks, you're so fast. Probably should've looked at clippy before pushing, fixing this and more now :)

**Review by osiewicz — APPROVED** [review]
LGTM, great job :)

**Comment by sven-frotscher** [comment]
Could `if __name__ == "__main__":` be made runnable in a similar way?
